python/ray/tune/impl/tuner_internal.py

```diff
@@ -123,6 +126,55 @@ def __init__(
         trainable_ckpt = os.path.join(self._experiment_checkpoint_dir, _TRAINABLE_PKL)
         with open(trainable_ckpt, "wb") as fp:
             pickle.dump(self._trainable, fp)
+        self._maybe_warn_resource_contention()
+
+    def _expected_utilization(self, cpus_per_trial, cpus_total):
+        num_samples = self._tune_config.num_samples
+        if num_samples < 0:  # TODO: simplify this in Tune
+            num_samples = math.inf
+        concurrent_trials = self._tune_config.max_concurrent_trials or 0
+        if concurrent_trials < 1:  # TODO: simplify this in Tune
+            concurrent_trials = math.inf
+
+        actual_concurrency = min(
+            (cpus_total // cpus_per_trial, num_samples, concurrent_trials)
+        )
+        return (actual_concurrency * cpus_per_trial) / cpus_total
+
+    def _maybe_warn_resource_contention(self):
+        trainable = self._convert_trainable(self._trainable)
+
+        # This may not be precise, but we don't have a great way of
+        # accessing the scaling config.
+        scaling_config = None
+        get_scaling_config = getattr(trainable, "base_scaling_config", None)
+        if callable(get_scaling_config):
+            scaling_config = get_scaling_config()
+
+        if not isinstance(scaling_config, ScalingConfig) or getattr(
+            scaling_config, "_max_cpu_fraction_per_node"
+        ):
+            return
+
+        has_dataset = getattr(trainable, "has_dataset", False)
```

Comment:
we really don't have a great abstraction for trainable for accessing parameters.